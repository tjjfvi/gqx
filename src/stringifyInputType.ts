
import { unwrapType } from "./unwrapType";
import { InputValueDefinitionNode } from "graphql";
import { Context } from "./stringifyOperations";
import { stringifyObject } from "./stringifyObject";

export const stringifyInputType = (ctx: Context, fields: InputValueDefinitionNode[]) =>
  stringifyObject(
    fields.map(field => {
      const [{ name: { value: typeName } }, wrap, maybe] = unwrapType(field.type, true);
      const type = (
        typeName in ctx.objectTypes || typeName in ctx.inputTypes || typeName in ctx.enumTypes ?
          typeName :
          `$$Scalars[${JSON.stringify(typeName)}]`
      );
      const typeStr = wrap(type);
      return [field.name.value, maybe, typeStr];
    })
  )
