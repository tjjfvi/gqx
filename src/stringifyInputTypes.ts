
import stringifyInputType from "./stringifyInputType";
import { Context } from ".";

export default (ctx: Context) =>
  Object.entries(ctx.inputTypes).map(([name, fields]) => {
    ctx.exports.push(name);
    return stringifyInputType(name, fields)
  }).join("\n\n")
