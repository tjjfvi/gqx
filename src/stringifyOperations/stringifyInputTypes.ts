
import { Obj } from ".";
import stringifyInputType from "../stringifyInputType";
import stringifyNamespace from "./stringifyNamespace";

export default (objs: Obj[]) =>
  stringifyNamespace("$$Input", objs, (o, p) => stringifyInputType(p.id.prop, p.args))
