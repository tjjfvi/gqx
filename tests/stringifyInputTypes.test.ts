
import stringifyInputType from "../src/stringifyInputType";
import stringifyInputTypes from "../src/stringifyInputTypes";

describe("stringifyInputTypes.gql", () => {
  const ctx = utils.getOperations(["stringifyInputTypes"]);
  test.each(Object.entries(ctx.inputTypes))("%s", (k, v) => {
    const result = stringifyInputType(k, v);
    expect(result).toParse().toMatchSnapshot();
  })
  test("*", () => {
    expect(stringifyInputTypes(ctx)).toParse().toMatchSnapshot();
  });
})
