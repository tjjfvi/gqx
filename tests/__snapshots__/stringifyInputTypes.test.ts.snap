// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`stringifyInputTypes.gql * 1`] = `
"export interface Mandatory {
  a: String;
  b: Int;
  c: Fraction;
}

export interface MandatoryList {
  a: String[];
  b: Int[];
  c: Fraction[];
}

export interface Nullable {
  a?: (String | null | void);
  b?: (Int | null | void);
  c?: (Fraction | null | void);
}

export interface NullableListOfMandatory {
  a?: (String[] | null | void);
  b?: (Int[] | null | void);
  c?: (Fraction[] | null | void);
}

export interface MandatoryListOfNullable {
  a: (String | null | void)[];
  b: (Int | null | void)[];
  c: (Fraction | null | void)[];
}

export interface NullableList {
  a?: ((String | null | void)[] | null | void);
  b?: ((Int | null | void)[] | null | void);
  c?: ((Fraction | null | void)[] | null | void);
}

export interface List2D {
  a: String[][];
  b: Int[][];
  c: Fraction[][];
}

export interface Fraction {
  numerator: Int;
  denominator: Int;
}"
`;

exports[`stringifyInputTypes.gql Fraction 1`] = `
"interface Fraction {
  numerator: Int;
  denominator: Int;
}"
`;

exports[`stringifyInputTypes.gql List2D 1`] = `
"interface List2D {
  a: String[][];
  b: Int[][];
  c: Fraction[][];
}"
`;

exports[`stringifyInputTypes.gql Mandatory 1`] = `
"interface Mandatory {
  a: String;
  b: Int;
  c: Fraction;
}"
`;

exports[`stringifyInputTypes.gql MandatoryList 1`] = `
"interface MandatoryList {
  a: String[];
  b: Int[];
  c: Fraction[];
}"
`;

exports[`stringifyInputTypes.gql MandatoryListOfNullable 1`] = `
"interface MandatoryListOfNullable {
  a: (String | null | void)[];
  b: (Int | null | void)[];
  c: (Fraction | null | void)[];
}"
`;

exports[`stringifyInputTypes.gql Nullable 1`] = `
"interface Nullable {
  a?: (String | null | void);
  b?: (Int | null | void);
  c?: (Fraction | null | void);
}"
`;

exports[`stringifyInputTypes.gql NullableList 1`] = `
"interface NullableList {
  a?: ((String | null | void)[] | null | void);
  b?: ((Int | null | void)[] | null | void);
  c?: ((Fraction | null | void)[] | null | void);
}"
`;

exports[`stringifyInputTypes.gql NullableListOfMandatory 1`] = `
"interface NullableListOfMandatory {
  a?: (String[] | null | void);
  b?: (Int[] | null | void);
  c?: (Fraction[] | null | void);
}"
`;
